server/src/main/java/io/crate/expression/predicate/NotPredicate.java

```diff
@@ -123,62 +110,57 @@ void add(Reference symbol) {
-                AnyRangeOperator.Comparison.LTE.opName(),
-                LikeOperators.ANY_LIKE,
-                LikeOperators.ANY_NOT_LIKE,
-                CoalesceFunction.NAME,
-                CaseFunction.NAME,
-                ConcatFunction.NAME,
-                ConcatWsFunction.NAME,
-                ArrayCatFunction.NAME,
-                ArrayAppendFunction.NAME,
-                ArrayFunction.NAME,
-                ArrayUniqueFunction.NAME,
-                FormatFunction.NAME,
-                IfFunction.NAME,
-                IsNullPredicate.NAME,
-                MapFunction.NAME,
-                TryCastFunction.NAME,
-                SubscriptObjectFunction.NAME
-            );
+        private final Set<String> CAST_FUNCTIONS = Set.of(ImplicitCastFunction.NAME, ExplicitCastFunction.NAME);
 
         @Override
-        public Void visitReference(Reference symbol, SymbolToNotNullContext context) {
+        public Void visitReference(Reference symbol, NullabilityContext context) {
             context.add(symbol);
             return null;
         }
 
         @Override
-        public Void visitFunction(Function function, SymbolToNotNullContext context) {
+        public Void visitFunction(Function function, NullabilityContext context) {
             String functionName = function.name();
-            if (Ignore3vlFunction.NAME.equals(functionName)) {
-                return null;
-            }
-            if (!STRICT_3VL_FUNCTIONS.contains(functionName)) {
-                for (Symbol arg : function.arguments()) {
-                    arg.accept(this, context);
+            if (CAST_FUNCTIONS.contains(functionName)) {
+                // Cast functions should be ignored except for the case where the incoming
+                // datatype is an object. There we need to exclude null values to not match
+                // empty objects on the query
+                var a = function.arguments().get(0);
+                var b = function.arguments().get(1);
+                if (a instanceof Reference ref && b instanceof Literal<?>) {
+                    if (ref.valueType().id() == DataTypes.UNTYPED_OBJECT.id()) {
+                        context.removeNullValues = true;
+                        return null;
+                    }
                 }
+            } else if (Ignore3vlFunction.NAME.equals(functionName)) {
+                context.useNotQuery = true;
+                return null;
             } else {
-                context.hasStrictThreeValuedLogicFunction = true;
+                var signature = function.signature();
+                if (signature.hasFeature(Feature.NULLABLE)) {
+                    context.removeNullValues = true;
+                } else if (signature.hasFeature(Feature.NON_NULLABLE)) {
+                    context.useNotQuery = true;
+                } else {
+                    // default case
+                    context.enforceThreeValuedLogic = true;
```

Comment:
If any function in the tree is true here, doesn't that mean that everything is true, exit early, and handle the logic differently below (`if (enforseThreeValueLogic) ... else if ... else`)?

Reply:
This is a good idea. This works.